### PR TITLE
Microsoft SQL Server data connector initial support

### PR DESCRIPTION
**Pull request by sgrebnov**
## 🗣 Description

Initial support for Microsoft SQL Server Data Connector.:
 - Auth/connection via `connection_string`
 - Schema inference (all data types except Timestampz and Binary types)
 - Query

```yaml
datasets:
  - from: mssql:saleslt.customer
    name: customer
    params: 
      mssql_connection_string: ${ secrets:MSSQL_CONNECTION_STRING }
```

The following will be done next:
 - More testing and improving error messages
 - Correct support for Timestampz and adding binary types support
 - Basic support for predicates pushdown
 - Streaming

## 🔨 Related Issues

Implements https://github.com/spiceai/spiceai/issues/2736
Part of https://github.com/spiceai/spiceai/issues/2688

## 🤔 Concerns

 Does not use `DbConnectionPool`, added minimal implementation as `SqlServerConnectionManager/SqlServerConnectionPool` directly to data component using bb8. This needs to be revised/improved.

### CODE REVIEW

File: crates/data_components/src/mssql.rs [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+    }
+
+    pub async fn get_schema(
+        conn: Arc<SqlServerConnectionPool>,
+        table: &TableReference,
+    ) -> Result<SchemaRef> {
+        let table_name = table.table();
+        let table_schema = table.schema().unwrap_or("dbo");
+
+        let columns_meta_query: String = format!(
```
Comment (sgrebnov) [inline crates/data_components/src/mssql.rs:115]: 'dbo' - dbo (Database Owner) is the default schema when a schema is not explicitly specified. If you create or access an object without specifying the schema, SQL Server will assume it belongs to the dbo schema

File: crates/data_components/src/mssql.rs [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+            table: table.clone(),
+        })
+    }
+
+    pub async fn get_schema(
+        conn: Arc<SqlServerConnectionPool>,
+        table: &TableReference,
+    ) -> Result<SchemaRef> {
+        let table_name = table.table();
+        let table_schema = table.schema().unwrap_or("dbo");
```
Comment (sgrebnov) [inline crates/data_components/src/mssql.rs:113]: TODO for @sgrebnov : use `catalog` (if provided). Could be added next

File: crates/data_components/src/mssql.rs [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+                    let (precision, scale) =
+                        get_column_precision_and_scale(column_name, schema).unwrap_or((38, 10));
+                    (Some(precision), Some(scale))
+                }
+                _ => (None, None),
+            };
+            let data_type =
+                map_column_type_to_arrow_type(column_type, decimal_precision, decimal_scale);
+
+            // arrow_fields.push(Field::new(column_name.clone(), data_type.clone(), true));
```
Comment (phillipleblanc) [inline crates/data_components/src/mssql.rs]:
```suggestion
```

File: crates/data_components/src/mssql.rs [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+                    let (precision, scale) =
+                        get_column_precision_and_scale(column_name, schema).unwrap_or((38, 10));
+                    (Some(precision), Some(scale))
+                }
+                _ => (None, None),
+            };
+            let data_type =
+                map_column_type_to_arrow_type(column_type, decimal_precision, decimal_scale);
+
+            arrow_columns_builders.push(map_data_type_to_array_builder_optional(Some(&data_type)));
```
Comment (phillipleblanc) [inline crates/data_components/src/mssql.rs]:
```suggestion
            arrow_columns_builders.push(map_data_type_to_array_builder(&data_type));
```
Can't we use the non-optional one here?

Comment (sgrebnov) [reply crates/data_components/src/mssql.rs]: Missed that we have non-optional one - updated. Thanks

File: crates/data_components/src/mssql.rs [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+        state: &dyn Session,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        let sql = format!("SELECT * from {dataset}", dataset = self.table);
+        let records = self.query_arrow(&sql, Arc::clone(&self.schema)).await?;
+
+        let table = MemTable::try_new(Arc::clone(&self.schema), vec![records])?;
+        table.scan(state, projection, filters, limit).await
```
Comment (phillipleblanc) [inline crates/data_components/src/mssql.rs:530]: The actual query shouldn't be done in the `scan` method. We need to implement an ExecutionPlan that defers the actual query to when .execute is called.

Comment (phillipleblanc) [reply crates/data_components/src/mssql.rs:530]: This can come in a follow-on PR

**Review by phillipleblanc — APPROVED** [review]

File: crates/data_components/src/mssql.rs (hunk 5 above)
Comment (sgrebnov) [reply crates/data_components/src/mssql.rs:530]: Confirm I'll do this as a next PR

**Pull request merged**